### PR TITLE
Fixed using float32 data

### PR DESCRIPTION
**Pull request by bkmi**
This allows for the use of float32,16,8,... data by means of casting up to float64 then casting back down to float32.

### CODE REVIEW

**Comment by L-sky** [comment]
@bkmi as for float32 specifically, script for calculating real spherical harmonics supports it natively (https://github.com/e3nn/e3nn/commit/54628d1f95641391f07634632165ab47ed97ef81) for some time now (though I recall some parts forced to float64 (CG?) that may cause type mismatch, and require type casting). 
Also, I don't see casting back, are those all the changes that were supposed to be committed?

File: e3nn/SO3.py [hunk 1]
```diff
@@ -314,14 +314,20 @@ def spherical_harmonics_xyz(order, xyz, sph_last=False, dtype=None, device=None)
...
+    elif xyz.dtype is not torch.float64:
+        xyz = xyz.to(dtype=torch.float64)
 
     with torch_default_dtype(torch.float64):
         if device.type == 'cuda' and max(order) <= 10:
             max_l = max(order)
-            out = xyz.new_empty(((max_l + 1)*(max_l + 1), xyz.size(0)))                                    # [filters, batch_size]
+            out = xyz.new_empty(
+                size=((max_l + 1)*(max_l + 1), xyz.size(0)),  # [ filters, batch_size]
+                dtype=torch.get_default_dtype()
```
Comment (L-sky) [inline e3nn/SO3.py]: it seems a bit excessive to have both this and typecast above, `new_empty` inherits dtype from `xyz`, so with mentioned typecast it is already float64

Comment (bkmi) [reply e3nn/SO3.py]: Agreed, it should be written only once.

**Comment by L-sky** [comment]
> When I try to use float32 data without this change cuda tries to access restricted memory.
Can you please send me inputs to the `rsh` that cause that?  

> The casting back is not a change. Check the return of the function.
Indeed, missed that part.

**Comment by bkmi** [comment]
Hmm actually I recompiled e3nn and I got it to work without the change.

**Pull request closed without merging**